### PR TITLE
DOC/CI: run doctests on travis

### PR DESCRIPTION
**Pull request by jorisvandenbossche**
Trying to see if we can check the doctests on travis. We were already running them, but at this point they were not failing travis if they failed.

Closes https://github.com/pandas-dev/pandas/issues/19934 
Closes https://github.com/pandas-dev/pandas/issues/3439

### CODE REVIEW

**Comment by TomAugspurger** [comment]
There were failures https://travis-ci.org/pandas-dev/pandas/jobs/347769808#L6516, but they build still passed.

**Comment by jorisvandenbossche** [comment]
Hmm, that is because the script is still called itself from another script that does `exit 0`. Added a commit to try to progagate this, but not sure that will work with this 'wait' stuff.

On the other hand, it is maybe better to move this to a separate script anyhow.

**Comment by jorisvandenbossche** [comment]
OK, this is working now, in the sense that travis is failing because of the failing doctests.

The question is now how to list the ones we want to test:

- manually list the files we want to test (and for some files, manually list the functions we want to test)
- or run all the doctests, and manually (for now) list the ones we want to skip

Checking how it would look like with the second way

**Comment by jorisvandenbossche** [comment]
Hmm, so running the doctests on travis gives segfaults for some of them:

```
...
pandas/core/frame.py::pandas.core.frame.DataFrame.count PASSED           [  4%]
pandas/core/frame.py::pandas.core.frame.DataFrame.cov PASSED             [  4%]
pandas/core/frame.py::pandas.core.frame.DataFrame.cummax *** Error in `/home/travis/miniconda3/envs/pandas/bin/python': munmap_chunk(): invalid pointer: 0x00000000027217b0 ***
ci/doctests.sh: line 57:  5790 Aborted                 (core dumped) pytest --doctest-modules --ignore=pandas/tests -v pandas
```

Somebody any idea?

**Comment by jorisvandenbossche** [comment]
So this still has a segfault when running the doctests.

I cannot reproduce it locally with pytest (`pytest --doctest-modules --ignore=pandas/tests -v pandas/core/frame.py`). 
But I actually can reproduce it when calling it more manually

```
import pandas as pd
import numpy as np
import doctest
doctest.testmod(pd.core.frame, globs={'pd': pd, 'np': np})
```

**Comment by jorisvandenbossche** [comment]
@jreback @TomAugspurger @datapythonista @WillAyd any comments? 

This enable doctests on travis, and for now runs tests for frame/generic/series.py. 
I took the approach of skipping those docstrings that don't work yet. We can then do follow-up PRs gradually fixing the others + removing its skip, but this at least ensures the others keep working or newly added will work.

File: ci/doctests.sh [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    echo "Running doctests"
+
+    # pytest --doctest-modules --ignore=pandas/tests -v  pandas/core/frame.py
+
+    # if [ $? -ne "0" ]; then
+    #     RET=1
+    # fi
+
+
+    # # top-level reshaping functions
```
Comment (jreback) [inline ci/doctests.sh]: are these just not-working tests?

**Review by jreback — APPROVED** [review]
small comments. generally +1 on this. Is there a doc about this in contributing? (certainly can do in later PRs), in particular formatting gotchas? Also list a couple of gold-star examples (in our current docs) that show good contstruction of these tests, so we can simply point to them and say, like this!

Once this is merged we can look to add these on new PRs and pls open a tracking issue for existing doc-tests to add to the tester

**Comment by datapythonista** [comment]
I think it'll be great to have this, but it'll take a while until we have all doctests passing (it's in my TODO list, together with fixing all warnings from sphinx).

I'm happy to merge this PR as it is, and keep adding docstrings and files as we fix the tests.

File: ci/doctests.sh (hunk 1 above)
Comment (jorisvandenbossche) [reply ci/doctests.sh]: Well, they are failing, and don't want to start fixing all of them here. Will remove or either skip the failing ones

**Comment by jorisvandenbossche** [comment]
> Is there a doc about this in contributing? (certainly can do in later PRs), in particular formatting gotchas? Also list a couple of gold-star examples (in our current docs) that show good contstruction of these tests, so we can simply point to them and say, like this!

Yes, we have a very elaborate docstring guide: http://pandas.pydata.org/pandas-docs/stable/contributing_docstring.html#docstring, which also contains some tips to get the doctests passing

(it might need some better linking in the contributing docs though)

File: doc/source/contributing.rst [hunk 2]
```diff
@@ -365,6 +365,31 @@ This will identify methods documented in ``doc/source/api.rst`` that are not act
...
+
+The examples in the docstring ('doctests') must be valid Python code,
+that in a deterministic way returns the presented output, and that can be
+copied and run by users. This can be checked with the script above, and is
+also tested on Travis. A failing doctest will be a blocker for merging a PR.
+Check the :ref:`examples <docstring.examples>` section in the docstring guide
+for some tips and tricks to get the doctests passing.
+
+When doing a PR with a docstring update, it is good to post the
+output of the validation script in a comment on github.
```
Comment (jorisvandenbossche) [inline doc/source/contributing.rst]: @datapythonista eventually we should rework the documentation guidelines to include more of sprint content like https://python-sprints.github.io/pandas/guide/pandas_pr.html, but for now added short section more clearly mentioning the validation script and that the doctests need to pass

File: doc/source/contributing.rst [hunk 3]
```diff
@@ -365,6 +365,31 @@ This will identify methods documented in ``doc/source/api.rst`` that are not act
 class methods, and existing methods that are not documented in ``doc/source/api.rst``.
 
 
+Updating a *pandas* docstring
+-----------------------------
+
+When improving a single function or method's docstring, it is not necessarily
+needed to build to full documentation (see next section).
```
Comment (datapythonista) [inline doc/source/contributing.rst]: s/to/the

**Review by datapythonista — APPROVED** [review]
Looks great, there is just a small typo.

**Review by TomAugspurger — APPROVED** [review]

**Pull request merged**